src/btreemap/node/io.rs

```diff
@@ -0,0 +1,363 @@
+            page_idx,
+            offset,
+            length,
+        } in iter
+        {
+            if page_idx == 0 {
+                // Write to the initial page.
+                write(
+                    self.allocator.memory(),
+                    (self.address + offset).get(),
+                    &src[bytes_written as usize..(bytes_written + length.get()) as usize],
+                );
+            } else {
+                // Write to an overflow page, allocating it if it doesn't already exist.
+                if self.overflows.len() < page_idx {
+                    self.allocate_new_page();
+                }
+
+                write(
+                    self.allocator.memory(),
+                    (self.overflows[page_idx - 1] + offset).get(),
+                    &src[bytes_written as usize..(bytes_written + length.get()) as usize],
+                );
+            }
+
+            bytes_written += length.get();
+        }
+    }
+
+    pub fn write_u32(&mut self, offset: Address, val: u32) {
+        self.write(offset, &val.to_le_bytes());
+    }
+
+    pub fn write_u64(&mut self, offset: Address, val: u64) {
+        self.write(offset, &val.to_le_bytes());
+    }
+
+    pub fn write_struct<T>(&mut self, t: &T, addr: Address) {
+        let slice = unsafe {
+            core::slice::from_raw_parts(t as *const _ as *const u8, core::mem::size_of::<T>())
+        };
+
+        self.write(addr, slice)
+    }
+
+    // Allocates a new page and appends it to the node's overflows.
+    fn allocate_new_page(&mut self) {
+        let new_page = self.allocator.allocate();
+        self.allocator
+            .memory()
+            .write(new_page.get(), &OVERFLOW_MAGIC[..]);
+        write_u64(
+            self.allocator.memory(),
+            new_page + PAGE_OVERFLOW_NEXT_OFFSET,
+            NULL.get(),
+        );
+
+        // Let the previous overflow page point to this one.
+        match self.overflows.last() {
+            Some(prev_overflow) => {
+                // There is a previous overflow. Update its next offset.
+                write_u64(
+                    self.allocator.memory(),
+                    *prev_overflow + PAGE_OVERFLOW_NEXT_OFFSET,
+                    new_page.get(),
+                );
+            }
+            None => {
+                // There is no previous overflow. Update the overflow address of the initial node.
+                write_u64(
+                    self.allocator.memory(),
+                    self.address + OVERFLOW_ADDRESS_OFFSET,
+                    new_page.get(),
+                );
+            }
+        }
+
+        self.overflows.push(new_page);
+    }
+
+    // Deallocates all the unused pages to avoid memory leaks.
+    fn deallocate_unused_pages(&mut self) {
+        // Compute how many overflow pages are needed.
+        let overflow_pages_needed = if self.max_offset as u32 > self.page_size.get() {
+            let overflow_page_capacity =
```

Comment:
would it be possible to move this logic to a separate small function and add unit tests?
I find this bit difficult to understand and to my experience those are the most common places for one-off errors.